private/bufpkg/bufprotoplugin/cmd/protoc-gen-go-apiclientconnect/main.go

```diff
@@ -0,0 +1,444 @@
+	g.P()
+
+	// WithAddressMapper functional option
+	g.P(`// WithAddressMapper maps the address with the given function.`)
+	g.P(`func WithAddressMapper(addressMapper func(string) string) ProviderOption {`)
+	g.P(`return func(provider *provider) {`)
+	g.P(`provider.addressMapper = addressMapper`)
+	g.P(`}`)
+	g.P(`}`)
+	g.P()
+
+	// WithContextModifierProvider functional option
+	// We create the contextModifier once for each address instead of passing address to a contextProvider directly
+	// so that we do not have to do the same address logic for each RPC call, and only do it when we create the provider.
+	// For example, we might read .netrc to create the contextProvider - we do not want to have to do this on every RPC call.
+	g.P(`// WithContextModifierProvider provides a function that  modifies the context before every RPC invocation.`)
+	g.P(`// Applied before the address mapper.`)
+	g.P(`func WithContextModifierProvider(contextModifierProvider func(address string) (func(`, contextGoIdentString, `) `, contextGoIdentString, `, error)) ProviderOption {`)
+	g.P(`return func(provider *provider) {`)
+	g.P(`provider.contextModifierProvider = contextModifierProvider`)
+	g.P(`}`)
+	g.P(`}`)
+	g.P()
+
+	// WithScheme functional option
+	g.P(`// WithScheme prepends the given scheme to the underlying transport address`)
+	g.P(`func WithScheme(scheme string) ProviderOption {`)
+	g.P(`return func(provider *provider) {`)
+	g.P(`provider.scheme = scheme`)
+	g.P(`}`)
+	g.P(`}`)
+	g.P()
+
+	// buildAddress helper function
+	g.P(`// buildAddress modifies the given address with any additional options for transport such as the scheme and any subdomains`)
+	g.P(`func (p *provider) buildAddress(address string) string {`)
+	g.P(`if p.addressMapper != nil {`)
+	g.P(`address = p.addressMapper(address)`)
+	g.P(`}`)
+	g.P(`if p.scheme != "" {`)
+	g.P(`address = p.scheme+"://"+address`)
+	g.P(`}`)
+	g.P(`return address`)
+	g.P(`}`)
+	g.P()
+
+	// Import path for the api named_go_package
+	apiGoImportPath, err := helper.NewPackageGoImportPath(
+		goPackageFileSet,
+		"api",
+	)
+	if err != nil {
+		return err
+	}
+	// Import path for the connectclient named_go_package
+	connectClientGoImportPath, err := helper.NewPackageGoImportPath(
+		goPackageFileSet,
+		"connectclient",
+	)
+	if err != nil {
+		return err
+	}
+
+	// Iterate over the services and create a constructor function for each
+	for _, service := range goPackageFileSet.Services() {
+		interfaceName := service.GoName
+		interfaceGoIdent := apiGoImportPath.Ident(interfaceName)
+		interfaceGoIdentString := g.QualifiedGoIdent(interfaceGoIdent)
+		interfaceConnectClientGoIdent := connectClientGoImportPath.Ident("New" + interfaceName + "Client")
+		interfaceConnectClientGoIdentString := g.QualifiedGoIdent(interfaceConnectClientGoIdent)
+
+		g.P(`func (p *provider) New`, interfaceName, `(ctx `, contextGoIdentString, `, baseURL string) (`, interfaceGoIdentString, `, error) {`)
```

Comment:
It seems that `context` isn't even needed here, but removing this doesn't make this quite a drop-in replacement as all the callsites will need changes to remove the context they are passing to `NewXXXXService`.  If this overall solution seems OK, though, I can remove it when the PR is submitted for final review.